morpheus-graphql-client/src/Data/Morpheus/Client/Fetch.hs

```diff
@@ -60,7 +60,8 @@ class Fetch a where
     where
       gqlReq = GQLRequest {operationName = Just opName, query = pack strQuery, variables = fixVars (toJSON vars)}
       -------------------------------------------------------------
-      processResponse JSONResponse {responseData = Just x} = Right x
+      processResponse JSONResponse {responseData = Just x, responseErrors = Nothing} = Right x
+      processResponse JSONResponse {responseData = Just x, responseErrors = Just []} = Right x
       processResponse invalidResponse = Left (show invalidResponse)
```

Comment:
what do you think, maybe we can use JSON "decode"  instead of regular show?

Reply:
will provide easier understandable error messages.

Reply:
So you might do something like:

```haskell
      processResponse JSONResponse {responseData = Just x, responseErrors = Nothing} = Right x
      processResponse JSONResponse {responseData = Just x, responseErrors = Just []} = Right x
      processResponse JSONResponse {responseErrors = err} = Left $ decodeUtf8 $ encode err
```

but you can't do

```haskell
     processResponse invalidResponse = Left $ decodeUtf8 $ encode invalidResponse
```

As `JSONResponse` has an `a` and based on the `Fetch` typeclass we don't know that we have `ToJSON a` constraint. So you would need to either add the `ToJSON a` constraint, which maybe the user doesn't want to provide. Or exclude the `responseData` from the error message.

Based on this issue I think the current implementation is fine. Also, I think unless you prettify the json, a json error message is not too different to a `show` error message. What do you think?

Reply:
If you wanted to improve this more, I think the right solution would be to have `fetch` return a `m (Either FetchFailure a)` where (roughly sketched)

```haskell
data FetchFailure = 
  FetchConnectionIssue String
  | ResponseDecodeFailure String
  | ServerGQLErrors (NonEmpty GQLError)
  deriving (Show, ToJSON)
```

This gives the user structured data they can work with if they want, or they can just show / render to JSON. I'd suggest this be done as a separate issue/PR though.